twitchio/ext/eventsub/websocket.py

```diff
@@ -0,0 +1,446 @@
+        if not self._subscription_pool:
+            return  # TODO: should this raise?
+
+        async with aiohttp.ClientSession() as session:
+            sock = self._sock = await session.ws_connect(reconnect_url or self.URL)
+            session.detach()
+
+        welcome = await sock.receive_json(loads=_loads, timeout=3)
+        logger.debug("Received websocket payload: %s", welcome)
+        self._session_id = welcome["payload"]["session"]["id"]
+        self._timeout = welcome["payload"]["session"]["keepalive_timeout_seconds"]
+
+        logger.debug("Created websocket connection with session ID: %s and timeout %s", self._session_id, self._timeout)
+
+        self._pump_task = self.client.loop.create_task(self.pump())
+
+        if reconnect_url:  # don't resubscribe to events
+            return
+
+        for sub in self._subscription_pool:
+            await self._subscribe(sub)  # TODO: how do I return this to the end user (do I bother?)
+
+    async def pump(self) -> None:
+        while self.is_connected:
+            try:
+                msg = await cast(aiohttp.ClientWebSocketResponse, self._sock).receive_str(
+                    timeout=self._timeout + 1
+                )  # extra jitter on the timeout in case of network lag
+                if not msg:
+                    continue  # TODO: should this raise?
+
+                logger.debug("Received websocket payload: %s", msg)
+                frame: _messages = self.parse_frame(_loads(msg))
+                self.client.run_event("eventsub_debug", frame)
+
+                if isinstance(frame, models.NotificationEvent):
+                    self.client.run_event(
+                        f"eventsub_notification_{models.SubscriptionTypes._name_map[frame.subscription.type]}", frame
+                    )
+                    self.client.run_event("eventsub_notification", frame)
+
+                elif isinstance(frame, models.RevokationEvent):
+                    self.client.run_event("eventsub_revokation", frame)
+
+                elif isinstance(frame, models.KeepAliveEvent):
+                    self.client.run_event("eventsub_keepalive", frame)
+
+                elif isinstance(frame, models.ReconnectEvent):
+                    self.client.run_event("eventsub_reconnect", frame)
+                    sock = self._sock
+                    self._sock = None
+                    await self.connect(frame.reconnect_url)
+                    await cast(aiohttp.ClientWebSocketResponse, sock).close(
+                        code=aiohttp.WSCloseCode.GOING_AWAY, message=b"reconnecting"
+                    )
+                    return
+
+            except asyncio.TimeoutError:
+                logger.warning(f"Websocket timed out (timeout: {self._timeout}), reconnecting")
+                await cast(aiohttp.ClientWebSocketResponse, self._sock).close(
+                    code=aiohttp.WSCloseCode.ABNORMAL_CLOSURE, message=b"timeout surpassed"
+                )
```

Comment:
Nitpick.

Whilst this code functions, it is very ugly and I would advise not making it a common pattern.
Casting should be done separately and be maintainably readable IMO.